### PR TITLE
Determine real, total, and cores for BSD

### PR DESCRIPTION
**Pull request by tas50**
Use dmesg to correctly determine the number of real CPUs, the number of
real cores, and the number of HT cores.  This aligns CPU count data with
what we collect on Linux.

### CODE REVIEW

**Comment by arachnist** [comment]
dmesg contents are volatile.
Relying on them is bound to produce errors or undefined behaviour sooner or later.

**Comment by kev009** [comment]
It should be fairly safe between majors

**Comment by lamont-granquist** [comment]
@arachnist i raised that question when we saw the same patch for dragonflybsd, but bsd users couldn't come up with a better API to query than dmesg...

File: lib/ohai/plugins/freebsd/cpu.rb [hunk 1]
```diff
@@ -54,13 +57,14 @@
         # Features2=0x80000001<SSE3,<b31>>
       when /Features2=[a-f\dx]+<(.+)>/
         cpuinfo["flags"].concat($1.downcase.split(','))
-      when /Logical CPUs per core: (\d+)/
-        cpuinfo["cores"] = $1
+      when /FreeBSD\/SMP: Multiprocessor System Detected: (\d*) CPUs/
+        cpuinfo["total"] = $1.to_i
+      when /FreeBSD\/SMP: (\d*) package\(s\) x (\d*) core\(s\)/
+        cpuinfo["real"] = $1.to_i
+        cpuinfo["cores"] = $2.to_i
```
Comment (lamont-granquist) [inline lib/ohai/plugins/freebsd/cpu.rb]: doesn't this need to be $1.to_i * $2.to_i?

Comment (tas50) [inline lib/ohai/plugins/freebsd/cpu.rb]: Ruby did the right thing, but I should add that incase it doesn't

Comment (tas50) [inline lib/ohai/plugins/freebsd/cpu.rb]: @lamont-granquist do you mean the total should be real * cores?  The line in dmesg above actually gives the number of cores including HT cores so we don't need to do the math.

Comment (lamont-granquist) [inline lib/ohai/plugins/freebsd/cpu.rb]: No, I'm talking about the 'cores' computation.  If this is a dual-socket quad-proc shouldn't that look like:

```
FreeBSD/SMP: Multiprocessor System Detected: 16 CPUs
FreeBSD/SMP: 2 package(s) x 4 core(s) x 2 SMT threads
```

and cpuinfo["cores"] should be 8 (2 x 4)

Comment (tas50) [inline lib/ohai/plugins/freebsd/cpu.rb]: yep.  let me fix that one too

**Comment by lamont-granquist** [comment]
:+1:

**Comment by mcquin** [comment]
LGTM

**Pull request merged**